migrator/src/main/scala/akka/persistence/jdbc/migrator/JournalMigrator.scala

```diff
@@ -0,0 +1,177 @@
+        // persist the data
+        writeJournalRows(row, tags)
+      }
+      .runWith(Sink.ignore)
+
+    // run the data migration and set the next ordering value
+    for {
+      _ <- eventualDone
+      _ <- journalOrdering.setSequenceVal()
+    } yield ()
+  }
+
+  /**
+   * Unpack a PersistentRepr into a PersistentRepr and set of tags.
+   * It returns a tuple containing the PersistentRepr and its set of tags
+   *
+   * @param pr the given PersistentRepr
+   */
+  private def unpackPersistentRepr(pr: PersistentRepr): (PersistentRepr, Set[String]) = {
+    pr.payload match {
+      case Tagged(payload, tags) => (pr.withPayload(payload), tags)
+      case _                     => (pr, Set.empty[String])
+    }
+  }
+
+  /**
+   * serialize the PersistentRepr and construct a JournalAkkaSerializationRow and set of matching tags
+   *
+   * @param pr       the PersistentRepr
+   * @param ordering the ordering of the PersistentRepr
+   * @return the tuple of JournalAkkaSerializationRow and set of tags
+   */
+  private def serialize(pr: PersistentRepr, ordering: Long): (JournalAkkaSerializationRow, Set[String]) = {
+
+    // let us unpack the PersistentRepr
+    val (repr, tags) = unpackPersistentRepr(pr)
+
+    val serializedPayload: AkkaSerialization.AkkaSerialized =
+      AkkaSerialization.serialize(serialization, repr.payload) match {
+        case Failure(exception) => throw exception
+        case Success(value)     => value
+      }
+
+    val serializedMetadata: Option[AkkaSerialization.AkkaSerialized] =
+      repr.metadata.flatMap(m => AkkaSerialization.serialize(serialization, m).toOption)
+    val row: JournalAkkaSerializationRow = JournalAkkaSerializationRow(
+      ordering,
+      repr.deleted,
+      repr.persistenceId,
+      repr.sequenceNr,
+      repr.writerUuid,
+      repr.timestamp,
+      repr.manifest,
+      serializedPayload.payload,
+      serializedPayload.serId,
+      serializedPayload.serManifest,
+      serializedMetadata.map(_.payload),
+      serializedMetadata.map(_.serId),
+      serializedMetadata.map(_.serManifest))
+
+    (row, tags)
+  }
+
+  /**
+   * inserts a serialized journal row with the mapping tags
+   *
+   * @param journalSerializedRow the serialized journal row
+   * @param tags                 the set of tags
+   */
+  private def writeJournalRows(journalSerializedRow: JournalAkkaSerializationRow, tags: Set[String]): Future[Unit] = {
+    val journalInsert: DBIO[Long] = newJournalQueries.JournalTable
+      .returning(newJournalQueries.JournalTable.map(_.ordering))
+      .forceInsert(
+        journalSerializedRow
+      ) // here we force the insertion of the auto-inc value to maintain the old ordering
+
+    val tagInserts: FixedSqlAction[Option[Int], NoStream, Effect.Write] =
+      newJournalQueries.TagTable ++= tags.map(tag => TagRow(journalSerializedRow.ordering, tag)).toSeq
+
+    journaldb.run(DBIO.seq(journalInsert, tagInserts).withPinnedSession.transactionally)
```

Comment:
have you tried to do the following? 

```scala
val combined = journalInsert.flatMap(_ => tagInserts)
journaldb.run(combined.transactionally)
```
I think the difference is that with `seq` one statement is executed against the DB and then other (two round trips). While with `flatMap` they are both sent to the DB as one command. But I'm not 100% sure about it. 

Also, we may not need a pinned session.